core/shared/src/main/scala/zio/ZQueue.scala

```diff
@@ -326,19 +333,58 @@ object ZQueue {
       p.unsafeDone(IO.succeedNow(a))
 
     sealed trait Strategy[A] {
-      def handleSurplus(as: List[A], queue: MutableConcurrentQueue[A], isShutdown: AtomicBoolean): UIO[Boolean]
+      def handleSurplus(
+        as: List[A],
+        queue: MutableConcurrentQueue[A],
+        takers: MutableConcurrentQueue[Promise[Nothing, A]],
+        isShutdown: AtomicBoolean
+      ): UIO[Boolean]
 
       def unsafeOnQueueEmptySpace(queue: MutableConcurrentQueue[A]): Unit
 
       def surplusSize: Int
 
       def shutdown: UIO[Unit]
+
+      @tailrec
+      final def unsafeCompleteTakers(
+        queue: MutableConcurrentQueue[A],
+        takers: MutableConcurrentQueue[Promise[Nothing, A]]
+      ): Unit =
+        pollTakersThenQueue(queue, takers) match {
+          case None =>
+          case Some((p, a)) =>
+            unsafeCompletePromise(p, a)
+            unsafeOnQueueEmptySpace(queue)
+            unsafeCompleteTakers(queue, takers)
+        }
+
+      @tailrec
+      private def pollTakersThenQueue(
```

Comment:
Inline into `unsafeCompleteTakers` to eliminate `Some` / `Tuple2` overhead.

Reply:
Done.